### ISSUE TITLE
Changing number of etcd nodes on a already running cluster

### ISSUE THREAD

**Issue by mumoshu** [kubernetes-incubator/kube-aws#647]
Extracted from https://github.com/kubernetes-incubator/kube-aws/issues/631#issue-227290960
The use-case is not supported as of today.

I'm not even aware of a possible implementation to achieve it but perhaps we can improve `etcdadm reconfigure` to detect an etcd cluster is in progress of adding a node to instruct etcd to dynamically join the existing cluster(by setting INITIAL_CLUSTER_STATE to `existing`). When in progress of removing a node, at least one of the existing nodes should run `etcdctl member remove` on the removed node so that the cluster removes it from the quorum.

**Comment by billyteves** [comment]
@mumoshu Is there a guide to letting us manually join the a new etcd member in the running cluster? I will be POCing using the the current ETCD CF in 0.9.6 then additionally, manually join it to the running cluster whenever it is possible once the cloud-init kicked off.

In addition to this, will this be the roadmap of ETCD? Will there be an ETCD cluster pods?

**Comment by mumoshu** [comment]
@billyteves Hi, thanks as always for trying kube-aws 👍 

It isn't on the roadmap mainly because I'm unsure why it is needed.

Changing number of etcd nodes is not for scaling-up an etcd cluster, but for more availability
However, then, why one needs to change number of etcd nodes after the initial startup?
Can't we just recreate the kube-aws cluster for such use-case?

Btw, scaling-up is achieved by upgrading the instance type of etcd nodes, which is already supported by kube-aws.

What's your use-case?

**Comment by mumoshu** [comment]
@billyteves Hi! Just checking in - any chance you could have some time to consider about my question?

**Comment by billyteves** [comment]
Hi @mumoshu apologies for so late reply, I've been busy for awhile. I think one of my use case is that,  other systems will be also running inside the cluster as a migration plan moving from traditional to cloud native. I have a assumption that we may need to increase the number of instances, not just vertically but also horizontally. At the same time, we may not destroy easily the cluster for reasons. But I would definitely agree with you in terms of planning out as early as possible. In any case, what would you recommend?

**Comment by mumoshu** [comment]
@billyteves Thanks for the reply!

> one of my use case is that, other systems will be also running inside the cluster as a migration plan moving from traditional to cloud native. 

I'm afraid I couldn't understand this part.
I'd rather like to connect between traditional apps hosted on VMs and cloud-native apps by deploying clusters for both kinds of apps into a single, shared VPC.
kube-aws has an ability to deploy k8s clusters into existing VPCs.
This way, traditional apps can communicate with cloud-native apps, and vice versa.

> I have a assumption that we may need to increase the number of instances, not just vertically but also horizontally.

Just curious but would you mind providing me a more concrete example of your use-case regarding the migration plan?
In which case while a migration, would you like to increase number of instances for etcd?

> At the same time, we may not destroy easily the cluster for reasons. 

Yes. However, you can bring up a brand-new cluster to gradually migrate pods from the old cluster to the new one. After all the pods have migrated to the new cluster, destroying the cluster won't be a huge problem. WDYT?

**Comment by whereisaaron** [comment]
> Btw, scaling-up is achieved by upgrading the instance type of etcd nodes, which is already supported by kube-aws.

Hi @mumoshu, I wasn't aware that changing the etcd instance type was possible. The [manual page](https://kubernetes-incubator.github.io/kube-aws/getting-started/step-4-update.html#the-etcd-caveat) seems to preclude this? You can do this with `kube-aws update`? Possibly [growing the disk volumes  or modifying the type or IOPS](http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-ec2-ebs-volume.html) is probably also possible via `kube-aws update`?

> There is no solution for hosting an etcd cluster in a way that is easily updateable in this fashion- so updates are automatically masked for the etcd instances. This means that, after the cluster is created, nothing about the etcd ec2 instances is allowed to be updated.

**Comment by mumoshu** [comment]
@whereisaaron Hi!

Honestly, I haven't tried it myself recently but I believe we can modify those via kube-update without any interruption now.

> EBS volumes support live configuration changes while in production. You can modify volume type, volume size, and IOPS capacity without service interruptions.
> ref: http://docs.aws.amazon.com/AWSEC2/latest/UserGuide/EBSVolumes.html

Also note that you can scale-up your Etcd nodes with no downtime - Etcd nodes are just replaced one-by-one with cfn-signal to achieve no downtime. This is supported since several months ago - I don't remember the exact kube-aws version introduced it though!

**Comment by whereisaaron** [comment]
Cool @mumoshu I knew EBS supported that, I've used it a lot, but I didn't know it wasn't masked out in `kube-aws`. I'll try to test this a bit and see how robust it is. The manual should probably be updated, noting at the least, the instance type change is possible.

**Comment by whereisaaron** [comment]
@mumoshu I just used `kube-aws upgrade` to change the instance type for a 3-node etcd cluster from t2.medium to m4.large. All went smoothly and a healthy cluster afterwards. Using kube-aws 0.9.9 with the `eni` identity method.

BTW, the eni method comments say:
>Use secondary ENIs and Route53 record sets to provide etcd nodes stable hostnames
I can see the 'identity' ENI interfaces, but I can't see any record set created in the managed Route53 zone?

**Comment by whereisaaron** [comment]
Using `kube-aws upgrade` to change the etcd root volume size worked for a three-node cluster, but only by terminating and recreating the instances and root volume. I think this is because the Launch Configuration was changed, which triggered complete replacement. The same etcd data volumes were retained and reattached appropriately (using `eni` identity method).

Using `kube-aws upgrade` to change the etcd data volume worked for a three-node cluster, the EBS volumes were dynamically enlarged (and then 'optimized' / copied) without affecting the running etcd nodes. I assume though, the file system will not actually enlarge unless I do that manually (or reboot the nodes?).

**Comment by whereisaaron** [comment]
While I was at it @mumoshu I tested deleting an etcd node from a three-node cluster (using ENI identity method), it was replaced without k8s API outage. I also tested simultaneously deleting two of the thee etcd nodes. There was a period of a minute or two where the k8s API servers didn't respond, but the remaining etcd node and two replacements quickly founds the S3 snapshot and recovered the three-node cluster based on that. Fantastic work @mumoshu 🎉 👍 👍 👍
